File: backend/services/comparison.py
from models.product import Product, ProductComparison


def compare_product(product: Product) -> ProductComparison:
    # Never compare different currencies or mix verified prices with demo prices.
    same_currency = [item for item in product.listings if item.currency == product.currency]
    source = "verified" if any(item.data_source == "verified" for item in same_currency) else "demo"
    eligible = [item for item in same_currency if item.data_source == source and item.availability == "available"
                and item.data_status not in {"STALE", "ERROR"}]
    lowest = min((item.price for item in eligible), default=None)
    highest = max((item.price for item in eligible), default=None)
    return ProductComparison(
        product=product,
        image_url=product.image_url,
        listings=product.listings,
        currency=product.currency,
        data_source=source if same_currency else "none",
        lowest_price=lowest,
        highest_price=highest,
        best_deals=[item for item in eligible if item.price == lowest],
        number_of_platforms=len({item.platform for item in product.listings}),
        last_updated=max((item.last_updated for item in product.listings), default=None),
    )

File: frontend/lib/comparison.ts
import type { Product } from "@/types/product";

export function compareProduct(product: Product) {
  const listings = product.listings ?? [];
  const sameCurrency = listings.filter(item => item.currency === product.currency);
  const source = sameCurrency.some(item => item.data_source === "verified") ? "verified" : "demo";
  const eligible = sameCurrency.filter(item => item.availability === "available" && item.data_source === source && item.data_status !== "STALE" && item.data_status !== "ERROR");
  return {
    listings, eligible, source,
    lowest: eligible.length ? Math.min(...eligible.map(item => item.price)) : null,
    highest: eligible.length ? Math.max(...eligible.map(item => item.price)) : null,
    platforms: new Set(listings.map(item => item.platform)).size,
    updatedAt: listings.length ? Math.max(...listings.map(item => Date.parse(item.last_updated))) : null,
  };
}

export function safeImageUrl(url?: string | null) {
  return url && (/^https?:\/\//i.test(url) || /^\/(?!\/)/.test(url)) ? url : undefined;
}

export function relativeTime(timestamp: number | null, now: number) {
  if (timestamp === null) return "Not yet collected";
  const seconds = Math.max(0, Math.floor((now - timestamp) / 1000));
  if (seconds < 60) return `${seconds}s ago`;
  if (seconds < 3600) return `${Math.floor(seconds / 60)} min ago`;
  if (seconds < 86400) return `${Math.floor(seconds / 3600)} hr ago`;
  return `${Math.floor(seconds / 86400)} days ago`;
}
